File: src/caa_scheduler/build_config.py
from __future__ import annotations

import re
from typing import Any


SHA256_PATTERN = re.compile(r"^[0-9a-f]{64}$", re.IGNORECASE)
AIRPORT_PATTERN = re.compile(r"^[A-Z0-9]{3,4}$")
FLEET_PATTERN = re.compile(r"^[A-Z0-9][A-Z0-9_-]*$")
VALID_MODES = {"mainline", "skunkworks"}
VALID_STARTING_POINTS = {"previous_schedule", "blank"}
VALID_NETWORK_ACTIONS = {"add", "remove", "status_change"}
VALID_CITY_STATUSES = {"destination", "focus_city", "hub", "inactive"}


def build_config_id(number: int, version: str) -> str:
    safe_version = re.sub(r"[^a-z0-9]+", "_", version.strip().lower()).strip("_")
    return f"schedule_{number}_v{safe_version or 'draft'}"


def _mapping(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _check(
    check_id: str, title: str, passed: bool, message: str
) -> dict[str, Any]:
    return {
        "id": check_id,
        "title": title,
        "status": "pass" if passed else "fail",
        "message": message,
    }


def validate_build_config(
    config: dict[str, Any], baseline: dict[str, Any] | None = None
) -> dict[str, Any]:
    """Validate the portable build contract before schedule construction starts."""
    if not isinstance(config, dict):
        return {
            "buildId": None,
            "status": "fail",
            "summary": {"checks": 1, "passed": 0, "failed": 1},
            "checks": [
                _check(
                    "document",
                    "Configuration document",
                    False,
                    "Build configuration must be a JSON object",
                )
            ],
        }
    checks: list[dict[str, Any]] = []
    checks.append(
        _check(
            "schema_version",
            "Configuration schema",
            config.get("schemaVersion") == "1.0.0",
            "Build configuration schema version is 1.0.0"
            if config.get("schemaVersion") == "1.0.0"
            else "Build configuration schemaVersion must be 1.0.0",
        )
    )
    schedule = _mapping(config.get("schedule"))
    number = schedule.get("number")
    version = schedule.get("version")
    label = schedule.get("label")
    identity_valid = (
        isinstance(number, int)
        and not isinstance(number, bool)
        and number > 0
        and isinstance(version, str)
        and bool(version.strip())
        and isinstance(label, str)
        and bool(label.strip())
    )
    checks.append(
        _check(
            "schedule_identity",
            "Schedule identity",
            identity_valid,
            f"Schedule {number}, version {version}, is explicitly identified"
            if identity_valid
            else "Schedule number, version, and label are required",
        )
    )

    expected_build_id = (
        build_config_id(number, version) if identity_valid else None
    )
    checks.append(
        _check(
            "build_id",
            "Build identifier",
            expected_build_id is not None and config.get("buildId") == expected_build_id,
            f"Build identifier is {expected_build_id}"
            if expected_build_id is not None
            and config.get("buildId") == expected_build_id
            else "buildId must be derived from the schedule number and version",
        )
    )

    mode = schedule.get("mode")
    checks.append(
        _check(
            "build_mode",
            "Build mode",
            mode in VALID_MODES,
            f"{mode} promotion rules will apply"
            if mode in VALID_MODES
            else "Build mode must be mainline or skunkworks",
        )
    )

    starting_point = _mapping(config.get("startingPoint"))
    start_kind = starting_point.get("kind")
    schedule_id = starting_point.get("scheduleId")
    start_valid = start_kind in VALID_STARTING_POINTS and (
        (start_kind == "blank" and schedule_id is None)
        or (
            start_kind == "previous_schedule"
            and isinstance(schedule_id, str)
            and bool(schedule_id)
            and (
                baseline is None
                or schedule_id == baseline.get("schedule", {}).get("id")
            )
        )
    )
    checks.append(
        _check(
            "starting_point",
            "Starting point",
            start_valid,
            "The starting point is explicit and matches the loaded baseline"
            if start_valid
            else "Choose a blank start or the loaded published baseline",
        )
    )

    fleet_counts = config.get("fleetCounts")
    fleet_valid = (
        isinstance(fleet_counts, dict)
        and bool(fleet_counts)
        and all(
            isinstance(fleet, str)
            and FLEET_PATTERN.fullmatch(fleet) is not None
            and isinstance(count, int)
            and not isinstance(count, bool)
            and count >= 0
            for fleet, count in fleet_counts.items()
        )
    )
    checks.append(
        _check(
            "fleet_plan",
            "Schedule-specific fleet plan",
            fleet_valid,
            (
                f"{len(fleet_counts)} fleet types and "
                f"{sum(fleet_counts.values())} aircraft are explicitly recorded"
            )
            if fleet_valid
            else "At least one fleet type with a nonnegative whole-aircraft count is required",
        )
    )

    connection = _mapping(config.get("connectionWindowMinutes"))
    minimum = connection.get("minimum")
    maximum = connection.get("maximum")
    connection_valid = (
        isinstance(minimum, int)
        and not isinstance(minimum, bool)
        and isinstance(maximum, int)
        and not isinstance(maximum, bool)
        and 0 <= minimum < maximum
    )
    checks.append(
        _check(
            "connection_window",
            "Connection window",
            connection_valid,
            f"Connections are pinned to {minimum}-{maximum} minutes"
            if connection_valid
            else "Connection bounds must be whole minutes with maximum greater than minimum",
        )
    )

    inputs = _mapping(config.get("inputs"))
    instructions = _mapping(inputs.get("instructions"))
    city_source = _mapping(inputs.get("cityInformation"))
    policy_source = _mapping(inputs.get("operatingPolicy"))
    demand_source = _mapping(inputs.get("demandData"))
    input_pins_valid = (
        bool(instructions.get("version"))
        and bool(instructions.get("source"))
        and bool(city_source.get("filename"))
        and SHA256_PATTERN.fullmatch(str(city_source.get("sha256", ""))) is not None
        and bool(policy_source.get("filename"))
        and SHA256_PATTERN.fullmatch(str(policy_source.get("sha256", ""))) is not None
        and bool(demand_source.get("version"))
    )
    checks.append(
        _check(
            "input_pins",
            "Pinned build inputs",
            input_pins_valid,
            "Instructions, city information, operating policy, and demand data are pinned"
            if input_pins_valid
            else "Instruction, city, policy, and demand pins must all be complete",
        )
    )

    pin_consistency = True
    pin_message = "Pinned city and operating-policy inputs match the loaded baseline"
    if (
        baseline is not None
        and input_pins_valid
        and start_kind == "previous_schedule"
    ):
        provenance = baseline.get("provenance", {})
        for key, configured in (
            ("cityInformation", city_source),
            ("operatingPolicy", policy_source),
        ):
            pinned = provenance.get(key, {})
            if (
                configured.get("filename") != pinned.get("filename")
                or str(configured.get("sha256", "")).lower()
                != str(pinned.get("sha256", "")).lower()
            ):
                pin_consistency = False
                pin_message = (
                    "This compiler can only seed from a baseline whose city and policy "
                    "fingerprints match the build configuration"
                )
                break
    checks.append(
        _check(
            "baseline_pin_consistency",
            "Baseline input consistency",
            pin_consistency,
            pin_message,
        )
    )

    active_cities = {
        str(city.get("code", "")).upper(): bool(city.get("active"))
        for city in (baseline or {}).get("cities", [])
    }
    known_cities = set(active_cities)
    changes = config.get("networkChanges", [])
    network_errors: list[str] = []
    seen: set[str] = set()
    if not isinstance(changes, list):
        network_errors.append("networkChanges must be a list")
        changes = []
    for index, change in enumerate(changes, start=1):
        if not isinstance(change, dict):
            network_errors.append(f"row {index}: change must be an object")
            continue
        airport = str(change.get("airport", "")).upper()
        action = change.get("action")
        target = change.get("targetStatus", "")
        if AIRPORT_PATTERN.fullmatch(airport) is None:
            network_errors.append(f"row {index}: invalid airport code")
        if airport in seen:
            network_errors.append(f"{airport or f'row {index}'}: duplicate change")
        seen.add(airport)
        if action not in VALID_NETWORK_ACTIONS:
            network_errors.append(f"{airport or f'row {index}'}: invalid action")
        if action == "add" and active_cities.get(airport):
            network_errors.append(f"{airport}: already active")
        if action == "remove" and baseline is not None and not active_cities.get(airport):
            network_errors.append(f"{airport}: not an active baseline city")
        if (
            action == "status_change"
            and baseline is not None
            and airport not in known_cities
        ):
            network_errors.append(f"{airport}: not present in the baseline city roster")
        if action in {"add", "status_change"} and target not in VALID_CITY_STATUSES:
            network_errors.append(f"{airport}: target status is required")
    checks.append(
        _check(
            "network_changes",
            "Network changes",
            not network_errors,
            f"{len(changes)} proposed airport changes passed structural preflight"
            if not network_errors
            else "; ".join(network_errors),
        )
    )

    policy_pinned = input_pins_valid and pin_consistency
    checks.append(
        _check(
            "curfew_enforcement",
            "Curfew enforcement",
            policy_pinned,
            "The pinned operating policy will enforce curfew violations as hard stops"
            if policy_pinned
            else "Curfew enforcement cannot be guaranteed until the operating policy is pinned",
        )
    )

    failed = sum(item["status"] == "fail" for item in checks)
    return {
        "buildId": config.get("buildId"),
        "status": "pass" if failed == 0 else "fail",
        "summary": {
            "checks": len(checks),
            "passed": len(checks) - failed,
            "failed": failed,
        },
        "checks": checks,
    }

File: web/build-config.mjs
export const BUILD_CONFIG_SCHEMA_VERSION = "1.0.0";

function clone(value) {
  return JSON.parse(JSON.stringify(value));
}

function cleanString(value) {
  return String(value ?? "").trim();
}

function positiveInteger(value) {
  const number = Number(value);
  return Number.isInteger(number) && number > 0 ? number : null;
}

function nonnegativeInteger(value) {
  const number = Number(value);
  return Number.isInteger(number) && number >= 0 ? number : null;
}

export function buildConfigId(number, version) {
  const safeVersion = cleanString(version).toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_|_$/g, "");
  return `schedule_${number}_v${safeVersion || "draft"}`;
}

export function createBuildConfig(canonical, manifest) {
  const source = canonical.schedule;
  const nextNumber = Number(source.number) + 1;
  const version = "0.1.0";
  return {
    schemaVersion: BUILD_CONFIG_SCHEMA_VERSION,
    buildId: buildConfigId(nextNumber, version),
    schedule: {
      number: nextNumber,
      version,
      label: `Schedule ${nextNumber}`,
      mode: "mainline",
      status: "draft",
    },
    startingPoint: {
      kind: "previous_schedule",
      scheduleId: source.id,
    },
    fleetCounts: clone(source.fleetCounts || {}),
    connectionWindowMinutes: clone(source.connectionWindowMinutes || { minimum: 30, maximum: 240 }),
    inputs: {
      instructions: {
        version: cleanString(manifest?.instructions?.version),
        source: cleanString(manifest?.instructions?.source),
      },
      cityInformation: clone(canonical.provenance?.cityInformation || { filename: "", sha256: "" }),
      operatingPolicy: clone(canonical.provenance?.operatingPolicy || { filename: "", sha256: "" }),
      demandData: {
        version: cleanString(manifest?.buildSetup?.demandData?.version),
      },
    },
    networkChanges: [],
    notes: "",
  };
}

export function normalizeBuildConfig(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError("Build configuration must be a JSON object.");
  }
  const schedule = value.schedule && typeof value.schedule === "object" ? value.schedule : {};
  const number = Number(schedule.number);
  const version = cleanString(schedule.version);
  const counts = value.fleetCounts && typeof value.fleetCounts === "object" && !Array.isArray(value.fleetCounts)
    ? Object.fromEntries(Object.entries(value.fleetCounts).map(([fleet, count]) => [cleanString(fleet).toUpperCase(), Number(count)]))
    : {};
  const normalized = {
    schemaVersion: cleanString(value.schemaVersion) || BUILD_CONFIG_SCHEMA_VERSION,
    buildId: buildConfigId(number, version),
    schedule: {
      number,
      version,
      label: cleanString(schedule.label),
      mode: cleanString(schedule.mode),
      status: "draft",
    },
    startingPoint: {
      kind: cleanString(value.startingPoint?.kind),
      scheduleId: value.startingPoint?.kind === "blank" ? null : cleanString(value.startingPoint?.scheduleId),
    },
    fleetCounts: counts,
    connectionWindowMinutes: {
      minimum: Number(value.connectionWindowMinutes?.minimum),
      maximum: Number(value.connectionWindowMinutes?.maximum),
    },
    inputs: {
      instructions: {
        version: cleanString(value.inputs?.instructions?.version),
        source: cleanString(value.inputs?.instructions?.source),
      },
      cityInformation: {
        filename: cleanString(value.inputs?.cityInformation?.filename),
        sha256: cleanString(value.inputs?.cityInformation?.sha256),
      },
      operatingPolicy: {
        filename: cleanString(value.inputs?.operatingPolicy?.filename),
        sha256: cleanString(value.inputs?.operatingPolicy?.sha256),
      },
      demandData: {
        version: cleanString(value.inputs?.demandData?.version),
      },
    },
    networkChanges: Array.isArray(value.networkChanges)
      ? value.networkChanges.map((change) => ({
          airport: cleanString(change?.airport).toUpperCase(),
          action: cleanString(change?.action),
          targetStatus: cleanString(change?.targetStatus),
          notes: cleanString(change?.notes),
        }))
      : [],
    notes: cleanString(value.notes),
  };
  return normalized;
}

function check(id, title, status, message) {
  return { id, title, status, message };
}

export function validateBuildConfig(value, canonical) {
  let config;
  try {
    config = normalizeBuildConfig(value);
  } catch (error) {
    return {
      status: "fail",
      summary: { checks: 1, passed: 0, blocking: 1, warnings: 0, notEvaluated: 0 },
      checks: [check("document", "Configuration document", "fail", error.message)],
    };
  }

  const checks = [];
  const scheduleNumber = positiveInteger(config.schedule.number);
  const identityValid = scheduleNumber !== null
    && Boolean(config.schedule.version)
    && Boolean(config.schedule.label);
  checks.push(check(
    "schedule_identity",
    "Schedule identity",
    identityValid ? "pass" : "fail",
    identityValid
      ? `Schedule ${scheduleNumber}, version ${config.schedule.version}, is explicitly identified.`
      : "Schedule number, version, and label are required; the number must be a positive integer.",
  ));

  const modeValid = ["mainline", "skunkworks"].includes(config.schedule.mode);
  checks.push(check(
    "build_mode",
    "Build mode",
    modeValid ? "pass" : "fail",
    modeValid
      ? `${config.schedule.mode === "mainline" ? "Mainline" : "Skunkworks"} promotion rules will apply.`
      : "Build mode must be Mainline or Skunkworks.",
  ));

  const baselineIds = new Set([canonical?.schedule?.id].filter(Boolean));
  const startValid = config.startingPoint.kind === "blank"
    || (
      config.startingPoint.kind === "previous_schedule"
      && Boolean(config.startingPoint.scheduleId)
      && baselineIds.has(config.startingPoint.scheduleId)
    );
  checks.push(check(
    "starting_point",
    "Starting point",
    startValid ? "pass" : "fail",
    startValid
      ? (config.startingPoint.kind === "blank"
          ? "The schedule will start from a blank network plan."
          : `The schedule is pinned to ${config.startingPoint.scheduleId}.`)
      : "Choose a blank start or a published baseline available in this console.",
  ));

  const fleets = Object.entries(config.fleetCounts);
  const invalidFleets = fleets.filter(([fleet, count]) => !fleet || nonnegativeInteger(count) === null);
  const duplicateNames = fleets.length !== new Set(fleets.map(([fleet]) => fleet)).size;
  const fleetValid = fleets.length > 0 && invalidFleets.length === 0 && !duplicateNames;
  checks.push(check(
    "fleet_plan",
    "Schedule-specific fleet plan",
    fleetValid ? "pass" : "fail",
    fleetValid
      ? `${fleets.length} fleet type${fleets.length === 1 ? "" : "s"} and ${fleets.reduce((sum, [, count]) => sum + Number(count), 0)} aircraft are explicitly recorded for this schedule.`
      : "Add at least one uniquely named fleet type and use whole-number aircraft counts of zero or greater.",
  ));

  const minimum = nonnegativeInteger(config.connectionWindowMinutes.minimum);
  const maximum = nonnegativeInteger(config.connectionWindowMinutes.maximum);
  const connectionValid = minimum !== null && maximum !== null && maximum > minimum;
  checks.push(check(
    "connection_window",
    "Connection window",
    connectionValid ? "pass" : "fail",
    connectionValid
      ? `Connections are pinned to ${minimum}–${maximum} minutes.`
      : "Connection minimum and maximum must be whole minutes, and maximum must exceed minimum.",
  ));

  const pins = config.inputs;
  const instructionPinned = Boolean(pins.instructions.version && pins.instructions.source);
  const cityPinned = Boolean(pins.cityInformation.filename && /^[0-9a-f]{64}$/i.test(pins.cityInformation.sha256));
  const policyPinned = Boolean(pins.operatingPolicy.filename && /^[0-9a-f]{64}$/i.test(pins.operatingPolicy.sha256));
  const demandPinned = Boolean(pins.demandData.version);
  const pinsValid = instructionPinned && cityPinned && policyPinned && demandPinned;
  checks.push(check(
    "input_pins",
    "Pinned build inputs",
    pinsValid ? "pass" : "fail",
    pinsValid
      ? "Instructions, city information, operating policy, and demand data are explicitly pinned."
      : "Pin the instruction version/source, city and policy files with SHA-256 fingerprints, and a demand-data version.",
  ));

  const activeCities = new Map(
    (canonical?.cities || []).map((city) => [String(city.code).toUpperCase(), Boolean(city.active)])
  );
  const allowedActions = new Set(["add", "remove", "status_change"]);
  const allowedStatuses = new Set(["destination", "focus_city", "hub", "inactive"]);
  const seenAirports = new Set();
  const networkErrors = [];
  config.networkChanges.forEach((change, index) => {
    const label = change.airport || `row ${index + 1}`;
    if (!/^[A-Z0-9]{3,4}$/.test(change.airport)) networkErrors.push(`${label}: invalid airport code`);
    if (!allowedActions.has(change.action)) networkErrors.push(`${label}: invalid action`);
    if (seenAirports.has(change.airport)) networkErrors.push(`${label}: duplicate change`);
    seenAirports.add(change.airport);
    if (change.action === "add" && activeCities.get(change.airport)) networkErrors.push(`${label}: already active`);
    if (change.action === "remove" && !activeCities.get(change.airport)) networkErrors.push(`${label}: not an active city`);
    if (["add", "status_change"].includes(change.action) && !allowedStatuses.has(change.targetStatus)) {
      networkErrors.push(`${label}: target status is required`);
    }
  });
  checks.push(check(
    "network_changes",
    "Network changes",
    networkErrors.length ? "fail" : "pass",
    networkErrors.length
      ? networkErrors.join("; ")
      : (config.networkChanges.length
          ? `${config.networkChanges.length} proposed airport change${config.networkChanges.length === 1 ? "" : "s"} passed structural preflight.`
          : "No airport additions, removals, or status changes are proposed."),
  ));

  const additions = config.networkChanges.filter((change) => change.action === "add").length;
  checks.push(check(
    "airport_metadata",
    "Airport metadata and runway screening",
    additions ? "warning" : "pass",
    additions
      ? `${additions} airport addition${additions === 1 ? "" : "s"} will require complete city metadata, demand inputs, and runway screening before construction.`
      : "No new airport metadata or runway screening is required at setup.",
  ));

  checks.push(check(
    "curfew_enforcement",
    "Curfew enforcement",
    policyPinned ? "pass" : "fail",
    policyPinned
      ? "The pinned operating policy marks curfew violations as construction hard stops."
      : "Curfew hard-stop enforcement cannot be guaranteed until the operating policy is pinned.",
  ));

  const summary = {
    checks: checks.length,
    passed: checks.filter((item) => item.status === "pass").length,
    blocking: checks.filter((item) => item.status === "fail").length,
    warnings: checks.filter((item) => item.status === "warning").length,
    notEvaluated: checks.filter((item) => item.status === "not_evaluated").length,
  };
  return {
    status: summary.blocking ? "fail" : "pass",
    summary,
    checks,
  };
}

export function serializeBuildConfig(value) {
  return `${JSON.stringify(normalizeBuildConfig(value), null, 2)}\n`;
}

export function buildConfigFilename(value) {
  const config = normalizeBuildConfig(value);
  return `${config.buildId}_build_config.json`;
}
